### PR TITLE
[ADD] web_view_leaflet_map (use leaflet.js library to have map views)

### PR DESCRIPTION
**Pull request by legalsylvain**
This PR introduces a new main module ``web_view_leaflet_map`` and a trivial module ``web_view_leaflet_map_partner`` that implements ``leaflet_map`` for ``res.partner`` model.

### Description

This module extends odoo views, to add a new kind of view, named ``leaflet_map`` that is using the Leaflet javascript library to use maps. (https://leafletjs.com/) This library is for exemple, used in the OpenStreetMap project. (https://www.openstreetmap.org/)

You can see a simple usage in the module ``web_view_leaflet_map_partner`` in the same OCA repository that displays your contact in a map, if latitude and longitude are defined. (To define latitude and longitude, refer to the Odoo module ``base_geolocalize``)
res_partner_map_1

res_partner_map_2

**Important Note**

The javascript library is opensource and distributed under BSD 2 Licence. See : https://github.com/Leaflet/Leaflet/blob/main/LICENSE. You can so use it freely.

However, display maps requires to display layers provided by tiles servers, that requires ressources.

**For testing purpose**

you can use the openStreetMap url ``https://tile.openstreetmap.org/${z}/${x}/${y}.png`` or other, listed in that page : https://wiki.openstreetmap.org/wiki/Tile_servers

Apart from very limited testing purposes, you should not use the tiles supplied by OpenStreetMap.org itself. OpenStreetMap is a volunteer-run non-profit body and cannot supply tiles for large-scale commercial use.

**Regular / High Usage**

- you can contact one of the following companies : https://switch2osm.org/providers/

- You can also install yourself your own tiles servers. See documentation : https://switch2osm.org/serving-tiles/

### Configure
* Go to Settings > Technical > Parameters > System Parameters

* Create or edit the parameter with the key ``leaflet.tile_url``

* As a value, set the url of the tiles server you chose. 

### Develop

Create a new view :

    <record id="view_my_model_map" model="ir.ui.view">
        <field name="model">my.model</field>
        <field name="arch" type="xml">
            <leaflet_map
                    field_latitude="FIELD_LATITUDE"
                    field_longitude="FIELD_LONGITUDE"
                    field_title="FIELD_TITLE"
                    field_address="FIELD_ADDRESS"
                    field_marker_icon_image="FIELD_MARKER_ICON_IMAGE"
                >
                <field name="__last_update"/>
                <field name="FIELD_LATITUDE"/>
                <field name="FIELD_LONGITUDE"/>
                <field name="FIELD_TITLE"/>
                <field name="FIELD_ADDRESS"/>
            </leaflet_map>
        </field>
    </record>

1. FIELD_LATITUDE and FIELD_LONGITUDE are the name of the fields that contains GPS coordinates of the model.
2. FIELD_TITLE will be used when the popup is displayed, as a title.
3. FIELD_ADDRESS will be used when the popup is displayed to display the adress.
4. (optional) FIELD_MARKER_ICON_IMAGE, is the name of the image field to place as an icon
   of the marker.
   Note: You can set extra settings ``marker_icon_size_x``, ``marker_icon_size_y``, to define
   the size of the image, and ``marker_popup_anchor_x``, ``marker_popup_anchor_y`` to define
   the position of the popup.

Map options :

- ``default_zoom`` : define the default zoom value. (7 if not defined)
- ``max_zoom`` : define the max zoom value. (19 if not defined)
- ``zoom_snap`` : define the zoom level in each change. (1 if not defined)

* Create or update an action for the model

        <record id="my_module.action_my_model" model="ir.actions.act_window">
            <field name="view_mode">tree,form,leaflet_map</field>
        </record>

**Library Update**

For the time being, the module embed the lealflet.js library version 1.8.0 ( released on April 18, 2022.)

If a new release is out:

- please download it here https://leafletjs.com/download.html
- update the javascript, css and images, present in the folder ``static/lib/leaflet``
- test the features
- make a Pull Request

**Default position in the map**

By default, the position of the map is defined by the user, in the function ``get_default_leaflet_position``. It returns the position of the current company, if defined. you can overload this function globally, or per model.

Note : Travis is broken on 12.0 branch. So CI is failing.

### CODE REVIEW

File: web_view_leaflet_map/static/src/js/view/map/map_renderer.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+    var MapRenderer = AbstractRenderer.extend({
+        tagName: 'table',
+
+        _render: function () {
+            console.debug("MapRenderer:_render");
+            var $mainDiv = $("<div id='div_map' class='_INTRODUCED_IN_RENDER'>");
+            this.$el.append($mainDiv);
+
+            const container = document.getElementById('div_map')
```
Comment (brian10048) [inline web_view_leaflet_map/static/src/js/view/map/map_renderer.js]: @legalsylvain could it be a missing semi-colon on L14?

Comment (legalsylvain) [reply web_view_leaflet_map/static/src/js/view/map/map_renderer.js]: Hi, @brian10048. Thanks for your answer. I added the semi-colon, unfortunately, it doesn't fix the error.

in fact, the rest of the code is executed, but container is null.

See : log

image

And result : 

image

Comment (brian10048) [reply web_view_leaflet_map/static/src/js/view/map/map_renderer.js]: I was just hoping for an easy answer!  Sorry I'm not really sure where to look next

File: web_view_leaflet_map/static/src/js/view/map/map_renderer.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+    var AbstractRenderer = require('web.AbstractRenderer');
+
+    var MapRenderer = AbstractRenderer.extend({
+        tagName: 'table',
+
+        _render: function () {
+            console.debug("MapRenderer:_render");
+            var $mainDiv = $("<div id='div_map' class='_INTRODUCED_IN_RENDER'>Content</div>");
+            this.$el.append($mainDiv);
```
Comment (fcayre) [inline web_view_leaflet_map/static/src/js/view/map/map_renderer.js]: Are you sure $el has already been added to the document at this time? If not, document.getElementById will not find it. What if you use container = $mainDiv[0] instead?

Comment (legalsylvain) [reply web_view_leaflet_map/static/src/js/view/map/map_renderer.js]: Thanks ! It worked like a charm. 

I have other problems. (see updated main comment), but PoC is OK !

**Comment by JulianaPoudou** [comment]
Hi @legalsylvain, 
I have a problem when I use leaflet lib with mobile, the map shows as grey (see file below):

Capture d’écran 2022-09-27 à 15 01 26

It seems that the css lines surrounded in red in file add the style transform "None" on image (from module web_editor) and blocks the map display.

Do you have the same behavior?
Thanks in advance!

**Comment by legalsylvain** [comment]
Hi. Thanks for testing @JulianaPoudou ! 

Indeed, just tested simulating mobile with firefox and I face the same issue.

Well, in fact i had the same problem in non mobile device, and i solved the problem with the call of 
``self.leaflet_map.invalidateSize();``

See : https://github.com/Leaflet/Leaflet/issues/3002#issuecomment-93836022

But I tried to change the timeout, and it does'nt work...
I don't know how to fix it. Any idea ? ;-)
 
Note : I will not investigate a lot, and I propose to add a point in fhe ``ROADMAP.rst`` file, because when we'll migrate the module in V16, we'll have to use owl2.

**Comment by legalsylvain** [comment]
Note : I took again a look on your remark. 
In "normal" view : here is the console : 

image

In mobile view : here is the console : 

image

AND, the _render function is not called. So I think it's the origin of the problem. However, I don't understand why the load of modules is failing in mobile mode. Any insight ?

Importante note : the bug is present is my module is not installed. it seems to be a bug in odoo core. 

CC : @Yajo @Tardo @SplashS  (maintainers of ``web_responsive`` module).

**Comment by SplashS** [comment]
> In mobile view : here is the console

Hello @legalsylvain 
It seems that you have problem with OdooTerminal. Please disable this plugin in browser and check again.
I tested your PR on runboat. Your control is working on my mobile iOS system.

**Comment by legalsylvain** [comment]
Hi @SplashS. 
I removed OdooTerminal. Indeed, it removed the error.
However, the map is bad displayed in mobile. Tested on Firefox 104.0 and Chromium Version 105.0.5195.125 (ubuntu 20.04)

**Comment by legalsylvain** [comment]
- CI is now green
- I added a ROADMAP section to talk about incompability with mobile device.

Could you update your review ? 

thanks !

**Comment by mathben** [comment]
Thanks for this PR <3

I test it and it works.

About the configuration, a problem occur, `https://tile.openstreetmap.org/${z}/${x}/${y}.png` not work, I use instead `https://tile.openstreetmap.org/{z}/{x}/{y}.png`.

About LeafletJS mobile, `self.leaflet_map.invalidateSize();` won't fix it, but this fix other problems.

Here, they use an iFrame to support mobile: https://leafletjs.com/examples/mobile/
Because, the modification will cause conflict with web (like disable unwanted scaling of the page).
To use iFrame, it needs some refactoring, like passing information between ir.ui.view and the iFrame.

**Comment by mathben** [comment]
Can we put the description of your PR into file `web_view_leaflet_map/README.rst`, because it missing documentation into the code how to configure it.

**Review by mathben — CHANGES_REQUESTED** [review]
Can we put the description of your PR into file `web_view_leaflet_map/README.rst`, because it missing documentation into the code how to configure it.
EDIT : the description is into `web_view_leaflet_map/readme/DESCRIPTION.rst`

File: web_view_leaflet_map/readme/DESCRIPTION.rst [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+The javascript library is opensource and distributed under BSD 2 Licence.
+See : https://github.com/Leaflet/Leaflet/blob/main/LICENSE
+You can so use it freely.
+
+However, display maps requires to display layers provided by tiles servers, that
+requires ressources.
+
+**For testing purpose**
+
+you can use the openStreetMap url ``https://tile.openstreetmap.org/${z}/${x}/${y}.png`` or other, listed in that page : https://wiki.openstreetmap.org/wiki/Tile_servers
```
Comment (mathben) [inline web_view_leaflet_map/readme/DESCRIPTION.rst]: Change ``https://tile.openstreetmap.org/${z}/${x}/${y}.png``
to `https://tile.openstreetmap.org/{z}/{x}/{y}.png`

**Review by mathben — CHANGES_REQUESTED** [review]

**Comment by legalsylvain** [comment]
hi @mathben. thanks for your review. i fixed the description. could you update the review ? 
BTW, could you also approve this trivial one ? https://github.com/OCA/geospatial/pull/319

**Review by mathben — APPROVED** [review]

**Review by Tardo — APPROVED** [review]
Tested on runbot and it seems to work correctly except for the problem with the mobile devices. Maybe to solve the issues with mobiles you can extend the leaflet library to append "!important" to the transform rule... or try to change the rule that cause problems in Odoo/Odoo :/

**Pull request merged**